pkg/client/client.go

```diff
@@ -1843,17 +1854,38 @@ func WithLastError(e *error) func(o *pollOptions) {
 // Poll is a wrapper around wait.PollUntilContextTimeout that allows adding the passed lastError into
 // the final error if set by the condition, adding more context to the "context deadline exceeded" error.
 func Poll(ctx context.Context, condition wait.ConditionWithContextFunc, options ...func(o *pollOptions)) error {
-	opts := pollOptions{timeout: 5 * time.Minute, interval: time.Second}
+	opts := pollOptions{
+		timeout:  5 * time.Minute,
+		interval: time.Second,
+	}
 	for _, o := range options {
 		o(&opts)
 	}
 
-	if err := wait.PollUntilContextTimeout(ctx, opts.interval, opts.timeout, false, condition); err != nil {
+	var (
+		conditionErr error
+		done         bool
+	)
+	deadlineCtx, deadlineCancel := context.WithTimeout(ctx, opts.timeout)
+	defer deadlineCancel()
+	if err := wait.PollUntilContextCancel(deadlineCtx, opts.interval, false, func(context.Context) (bool, error) {
+		// Don't pass the deadline context to the condition function to avoid
+		// errors when calling the API server.
+		done, conditionErr = condition(ctx)
```

Comment:
I'm not sure I fully understand. Iiuc the goal here is to have `func Poll` be in charge of the timeout/deadline for the polling, instead of sharing the deadline ctx with the `condition()`?
If that is right, shouldn't we pass a completely new context to `condition()`? With the current implementation a caller of `Poll()` can still pass in a ctx with a timeout and our new `deadlineCtx` can only ever timeout earlier, but no later the the passed ctx.

Reply:
Yes, the idea is that the condition function should complete even if the poll context terminates during its execution.

In my first version, I created a new context but then I realize that the parent context is a cancel-only context which is done only when the program terminates (SIGTERM) so it feels better to stop the condition function ASAP in this case.

Reply:
Oh I see now. `PollUntilContextTimeout` sets the timeout on `ctx` before passing it to `condition`. Thanks